### DISCUSSION TITLE
CloudFlare R2 - Can't Establish Connection

### DISCUSSION THREAD

**Discussion by chrisspiegl** [iterate-ch/cyberduck#16027] [Q&A]
I am trying to use R2 with Mountainduck.
I set up the API Access Key ID and Secret Access Key via the R2 interface and got the URL for the bucket.
Separating the bucket name `/files-production` from the URL and entered it with the `Cloudflare R2 Storage (S3)` Profile.

However, I can't establish a connection and wonder what may be the case? I tried looking at the [docs](https://docs.cyberduck.io/protocols/s3/cloudflare/) but could not get any additional information.

I also noticed that the `<key>Hostname Placeholder</key><string>&lt;accountid&gt;.r2.cloudflarestorage.com</string>` in the profile does not appear to be working since instead of this, it just displays the `s3.amazonaws.com` in the `Server` field. Or rather: the placeholder is working, but the default URL that is visible when selecting the profile is `s3.amazonaws.com`.

Anyone with a working R2 connection? Maybe you can help me out here?

Here is the login error message and the config of the R2 API Keys:

<img width="368" alt="Xnapper-2024-06-14-10 01 49" src="https://github.com/iterate-ch/cyberduck/assets/1096384/70c7e939-a32d-4eef-ae75-5aa74fb5c410">

<img width="1038" alt="Xnapper-2024-06-14-10 05 34" src="https://github.com/iterate-ch/cyberduck/assets/1096384/988d17d6-2a84-4448-8ddd-68948ba72da6">

**Comment by dkocher** [comment]
Please confirm you are using the _Cloudflare R2 Storage (S3)_ enabled from _Preferences → Profiles_.

**Reply by chrisspiegl** [reply]
Yes, that is undoubtedly the one I chose.
I first selected it from the Preferences → Profiles, then also tried downloading it from the profiles repo to open it and make sure it's the right one.

I can also confirm that when the `Server` field is empty, the placeholder does show. I just was confused that it was pre-filled with `s3.amazonaws.com` instead of a `xxx.r2.cloudflarestorage.com` type domain.

**Reply by dkocher** [reply]
Can you add a screenshot from your bookmark window.

**Reply by chrisspiegl** [reply]
Here is a screenshot of the bookmark window:

<img width="795" alt="Xnapper-2024-06-15-13 16 52" src="https://github.com/iterate-ch/cyberduck/assets/1096384/435754c4-e8ff-47cc-b3d7-be12b1816327">

**Reply by dkocher** [reply]
Can you confirm you have only the Account ID in the hostname set in _Server_ but **not** the bucket. I cannot reproduce with a equivalent setup accessing a single bucket in CloudFlare R2 by setting it in the _Path_ field.

**Reply by chrisspiegl** [reply]
I have set the details exactly as shown on the screenshot.
In the `Server:` field, I have `ACCOUNTID.r2.cloudfalrestoreage.com` as it is given by Cloudfalre.
I cannot enter any path after the `.com` since then an error `DNS LOOKUP FAILED` will happen.

The `Path:` field is then set to the bucket name based on how it is set up on CloudFlare.

I also have worked with Cloudflare in other S3 libraries when working directly with code, and it did work. So I am not clear what may be going on. Which is why I am reaching out for help here. Thinking I may be missing something? Some setting I should chagne? Set different?

**Reply by dkocher** [reply]
Your configuration looks sane. Do you connect through a HTTP proxy that might interfere with request headers?

**Reply by chrisspiegl** [reply]
No HTTP Proxy as far as I am aware. That is very strange.

**Reply by chrisspiegl** [reply]
I had time to test some more and found out the following (which finalizes this for me for now):

My initial testing with my `files-production` bucket kept not working. This bucket was set to be located in the EU based on the ability to set this in Cloudflare when creating the bucket. Then you also use the `ACCOUNT_ID.eu.r2.cloudflarestorage.com` URL (note `.eu.` in there).

Then I created a test bucket which was not “locked” to be in the EU, and I was able to connect to that bucket (obviously also using the `ACCOUNT_ID.r2.cloudflarestorage.com` the version without the `.eu.`).

I then created a new bucket which has a “location hint” to be set to “Western Europe”. Again, it is working (without the `.eu.`).

**So, using “Location: Automatic” instead of “Location: Specify jurisdiction” works.**

I don't quite understand why that may be… and I don't care enough about the “Specific jurisdiction” to investigate further, now that I have a working version.

<img width="935" alt="Xnapper-2024-06-18-11 30 27" src="https://github.com/iterate-ch/cyberduck/assets/1096384/31e6ee44-b2ae-4790-a5d1-475d05921c15">

**Comment by dkocher** [comment]
It looks like CloudFlare R2 introduced support for location constraints [^1]. I recall previously only `auto` was available and is set accordingly in the connection profile. Because the region is part of the request signature calculation that explains the failure authenticating for buckets with a location constraints. I have opened [^2] to look into this.

[^1]: https://developers.cloudflare.com/r2/reference/data-location/#available-hints
[^2]: iterate-ch/profiles#91

**Reply by chrisspiegl** [reply]
That explains a lot. One thing to note: it's not just about the `available-hints` but more importantly, in my case, about the [Jurisdictions with S3 API](https://developers.cloudflare.com/r2/reference/data-location/#using-jurisdictions-with-the-s3-api).

The strange thing here (based on the docs by Cloudflare): 

They say that the `region: "auto"` should still be set to auto and the only difference is the url containing the `.eu.` qualifier?

**Reply by dkocher** [reply]
The use of a different endpoint only applies to the jurisdictional feature. But I assume the same applies here that requests must be signed for the specific location instead of `auto`.
> Jurisdictional buckets can only be accessed via the corresponding jurisdictional endpoint. [^1] 

[^1]: https://developers.cloudflare.com/r2/api/s3/tokens/

**Reply by dkocher** [reply]
@chrisspiegl You may want to try the connection profile from [^1] by replacing the version in the _Profiles_ of your [application support folder](https://docs.cyberduck.io/cyberduck/support/#application-support-folder).

[^1]: iterate-ch/profiles#92

**Reply by chrisspiegl** [reply]
I'll try the profile; however, I think these are two different things.

From the Cloudflare Docs, they literally show in the code example that one should use `region: "auto"` when using the jurisdictional buckets and only change the `corresponding endpoint`? Which makes me believe that the “region hints” have nothing to-do with it (see code example from the Cloudflare docs page):

<img width="851" alt="Xnapper-2024-06-18-16 54 24" src="https://github.com/iterate-ch/cyberduck/assets/1096384/9606f5aa-1f58-460a-ac6b-2c04af229d46">

**Reply by varundeva** [reply]
I'm also facing the same issue. 
did u find any solution?

**Reply by chrisspiegl** [reply]
So far, I think it's still the same. Connecting to a `eu` location locked `r2` bucket is not working for some reason. But the global ones (without `.eu.` in the URL) should work. As disused here https://github.com/iterate-ch/cyberduck/discussions/16027#discussioncomment-9804209
